### ISSUE TITLE
jacoco-maven-plugin changes behaviour of the nameservice service locator

### ISSUE THREAD

**Issue by janotav** [jacoco/jacoco#591]
### Steps to reproduce

My application uses the _-Dsun.net.spi.nameservice.provider.1_ system property in order load custom name service implementation. 

After enabling jacoco to gather coverage, the mechanism for selecting provider stops working and default (system) implementation is simply used.

The following maven project demonstrates the issue:
[jacocoAndDns.zip](https://github.com/jacoco/jacoco/files/1288190/jacocoAndDns.zip)

When jacoco is not involved (mvn clean install), the output is:
www.google.com: 127.0.0.1
www.amazon.com: 127.0.0.1

With jacoco (mvn clean install -Pjacoco) my dummy nameservice implementation is not loaded and regular resolver is used:
www.google.com: 216.58.204.68
www.amazon.com: 52.84.121.20

JaCoCo version: jacoco-maven-plugin 0.7.9 
Operating system: Linux

### Expected behaviour
"prepare-agent" should not have impact on the Java APIs

### Actual behaviour
_-Dsun.net.spi.nameservice.provider.1_ property is ignored when jacoco agent is loaded

**Comment by Godin** [comment]
Hi @janotav ,

First of all - thank you for this concise example of your problem! 👍

`prepare-agent` goal of `jacoco-maven-plugin` simply sets property `argLine` used by `maven-surefire-plugin` into something like `-javaagent:path/to/jacoco/agent.jar=arguments` ( http://www.jacoco.org/jacoco/trunk/doc/prepare-agent-mojo.html ) , what results in passing of this value on command-line as `java -javaagent:...`

So let's simplify your example further down to the following two files:

first - combination of your `Hello.java` with `HelloTest.java` into `TestEnvNameServic.java`:
```
import sun.net.spi.nameservice.NameService;
import sun.net.spi.nameservice.NameServiceDescriptor;

import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.UnknownHostException;

public class TestEnvNameService implements NameServiceDescriptor {
    public static void main(String[] args) throws Exception {
        resolve("www.google.com");
    }

    private static void resolve(String name) throws UnknownHostException {
        InetAddress[] address = InetAddress.getAllByName(name);
        System.out.println(name + ": " + address[0].getHostAddress());
    }

    @Override
    public NameService createNameService() throws Exception {
        return new NameService() {
            @Override
            public InetAddress[] lookupAllHostAddr(String hostname) throws UnknownHostException {
                return new InetAddress[] { Inet4Address.getByAddress(hostname, new byte[] { 127, 0, 0, 1 }) };
            }

            @Override
            public String getHostByAddr(byte[] bytes) throws UnknownHostException {
                throw new UnknownHostException();
            }
        };
    }

    @Override
    public String getProviderName() {
        return "testDns";
    }

    @Override
    public String getType() {
        return "dns";
    }
}
```

second - `META-INF/services/sun.net.spi.nameservice.NameServiceDescriptor`:
```
TestEnvNameService
```

And execute
```
javac TestEnvNameService.java
jar cf test.jar META-INF *.class
java \
    -cp test.jar \
    -Dsun.net.spi.nameservice.provider.1=dns,testDns \
    -javaagent:org.jacoco.agent-0.7.9-runtime.jar \
    TestEnvNameService
```
which shows
```
www.google.com: 127.0.0.1
```

so that statement

> -Dsun.net.spi.nameservice.provider.1 property is ignored when jacoco agent is loaded

is not true.

But then fair question - why original example behaves differently? The answer is "hidden" (not very obvious) in `maven-surefire-plugin` documentation in the last paragraph of page http://maven.apache.org/surefire/maven-surefire-plugin/examples/system-properties.html :
> Special VM Properties
> Some system properties must be set on the command line of the forked VM, and cannot be set after the VM has been started. These properties must be added to the argLine parameter of the Surefire plugin. 

Rephrasing: system properties specified via `<systemProperties>` of `maven-surefire-plugin` are set after VM startup, because not passed on command line - see https://github.com/apache/maven-surefire/blob/surefire-2.20/surefire-booter/src/main/java/org/apache/maven/surefire/booter/PropertiesWrapper.java#L50-L56

Strictly speaking there is no guarantees that setting of property `sun.net.spi.nameservice.provider.1` after JVM startup will have effect.

On practice - setting of this property after it has been read won't have effect.

Startup of any Java agent happens after JVM startup and before execution of `main`. I.e. "surefire booter" is executed after startup of JaCoCo agent and so after following code that causes reading of property: https://github.com/jacoco/jacoco/blob/4b471ef7bb7992a6fb213a2e3e4ec3e03c9d02e5/org.jacoco.agent.rt/src/org/jacoco/agent/rt/internal/Agent.java#L169-L172

To sum up - solutions:
* specify `sun.net.spi.nameservice.provider.1` using `argLine`
* specify `sessionId` to prevent its generation, but keep in mind that there is no guarantees that setting of `sun.net.spi.nameservice.provider.1` after JVM startup will work, e.g. that JVM itself or one of its components (such as JMX) won't read it during its startup

Given existence of solutions and mainly nature of root cause of problem, I don't think that this should be classified as JaCoCo "bug".

Given that algorithm of generation of random `sessionId` is an implementation detail, we might change it as low-priority "enhancement".

**Comment by janotav** [comment]
Hi @Godin,

Thank you for your detailed analysis. In my particular case simply switching to the argLine mechanism properly fixes the problem (I had a direct InetAddress manipulation workaround in place before). 

Thanks.
